Async-http server.

HelenMamontova opened the pull request:
Создан асинхронный HTTP сервер.

HelenMamontova committed:
The .gitignore new file added.

HelenMamontova committed:
The server.cpp new file with <boost/asio.hpp>, <iostream>, <string>, <ctime>
header files and make_daytime_string() function added.

HelenMamontova committed:
The Makefile added.

HelenMamontova committed:
The *.o, *.d, *.txt, *.html, *.htm, server_files lines added to
.gitignore file.

HelenMamontova committed:
The reference() function added to server.cpp file.

HelenMamontova committed:
The main() function with parsing command line parameters
and defining variables port and host added.

HelenMamontova committed:
The io_service object end run() function added in the main() function.

HelenMamontova committed:
The tcp_server class added. The tcp_server server object added in main()
function.

HelenMamontova committed:
The host and port parameters, reslver.async_resolve() call added in the
tcp_server() constructor. The handle_reslve() function added in the class
tcp_server.

HelenMamontova committed:
Initialization of class member acceptor_ added to constructor tcp_server().
The if(!err)-else construction added to handle_resolve() function.

HelenMamontova committed:
The socket_ member of class tcp_server added and initialized.
The function handle_resolve() using error code commented out.
The handle_reslve() function using exceptions added.
The handle_accept() functin added.

HelenMamontova committed:
The tcp::endpint ep object moved inside the while cycle,
the acceptor_.open() method call moved inside the while cycle and try block,
the acceptor_.close() method call added in the handle_resolve() method.

HelenMamontova committed:
Header file <boost/shared_ptr.hpp> added.
The tcp_server class member socket_ removed.
Commented code removed.
The socket object added to the handle_reslve() function.

HelenMamontova committed:
The tcp_server class member boost::asio::io_service& io_service_ object added.
In the handle_resolve() function the object io_service deleted,
and when creating new tcp::socket() the parameter io_service replaced by the parameter io_service_.

HelenMamontova committed:
The handle_write() function added. The start_accept() function added.
The function call async_write() and function call make_daytime_string() added to function handle_accept().

HelenMamontova committed:
Function make_log_line() added. Variable message changed to mes,
log message for stdout changed in function handle_accept().
Version fixed to 1.0.0 in main() function.

HelenMamontova committed:
Header file <fstream> added.
A class member variable outfile_ and its initialization in the constructor tcp_server() added.
The outfile parameter in tcp_server() constructor added.
The write_log() method and its call in handle_accept() function added.

HelenMamontova committed:
Parameter boost::asio::transfer_all for function async_write() added in handle_write() function.
Parameter 'bytes_transfered' in function handle_write() commented out.

HelenMamontova committed:
tcpserver.cpp and tcpserver.o added in Makefile.

HelenMamontova committed:
The new file tcpserver.h added.

HelenMamontova committed:
The new file tcpserver.cpp added.

HelenMamontova committed:
Class tcp_server renamed to Tcpserver and moved to separate files tcpserver.h, tcpserver.cpp.

HelenMamontova committed:
The header file "request.h" added Formatting fixed.

HelenMamontova committed:
The new file request.h added.

HelenMamontova committed:
The new file request.cpp added.

HelenMamontova committed:
The tcpconnection.h new file added.

HelenMamontova committed:
The tcpconnection.cpp new file added.

HelenMamontova committed:
tcpconnection.cpp, tcpconnection.o added in the Makefile.

HelenMamontova committed:
The 'using' and 'typedef' directives removed in the server.cpp file.

HelenMamontova committed:
The header file "tcpconnection" added. The handle_write() method
removed. The socket parameter replaced by connection in the
handle_accept() method. The class member variable char buff added.

HelenMamontova committed:
The header file "tcpconnection.h" added. The handle_write() method removed. The socket parameter replaced by connection in the handle_accept(), async_accept() functions.

HelenMamontova committed:
The read_complete() method added. The bytes_transfered parameter removed in handle_read() method.

HelenMamontova committed:
The read_complete() method added. The bytes_transfered parameter removed in handle_read() method.
Getting the value of the variable message_ and checking if(bytes < 1024) added in the handle_read() method.

HelenMamontova committed:
The boost::shared_ptr parameter changed to std::shared_ptr in the
handle_accept() method.

HelenMamontova committed:
The typedef boost::shared_ptr changed to typedef std::shared_ptr.

HelenMamontova committed:
Added inheritance of Tcpconnection class from
std::enable_shared_from_this class.

HelenMamontova committed:
The namespace pls added. Header file <boost/bind.hpp> removed. Boost::bind
changed to std::bind in the parameters of the async_read() function.

HelenMamontova committed:
The async_read() function call added if bytes = 1024 in the
handle-read() method.

HelenMamontova committed:
request.cpp, request.o added in the Makefile.

HelenMamontova committed:
The 'work_dir' parameter added when Tcpserver server() call.

HelenMamontova committed:
The std::string& work_dir added in the Tcpserver() constructor. The
std::string work_dir_ member of Tcpserver class added. The std::string&
outfile_ member of Tcpserver class replaced by std::string outfile_.

HelenMamontova committed:
Class member initialization work_dir_ added to constructor Tcpserver().
The work_dir_ parameter added when cnnection_ptr connection() call.

HelenMamontova committed:
The header files "request.h", <dirent.h> added. The class member
std::string work_dir_ added. The work_dir parameter added in the
Tcpconnection() constructor. The make_index(), make_response(),
handle_read() methods added.

HelenMamontova committed:
Class member initialization work_dir_ added to constructor Tcpconnection().
The make_index(), make_response() methods added. The make_response()
method call added in the handle_read() method.

HelenMamontova committed:
The handle_write() method declaration added.

HelenMamontova committed:
The handle_write() methd definition added. Formatting fixed.

HelenMamontova committed:
The string_to_vector_char() method added. The type of make_index(),
make_response() methods changed to std::vector<char>.

HelenMamontova committed:
The string_to_vector_char() method definition added. The type of make_index(),
make_response() methods changed to std::vector<char>. The method call string_to_vector_char()  for strings added.
The header files <fstream>, <vector>, <algorithm>, <cerrno>, <dirent.h>, <sys/stat.h> added.
The block if(request.path()!="/") added in make_response() method.

HelenMamontova committed:
The server.cpp file renamed to main.cpp.

HelenMamontova committed:
The string "server" changed to "main".

HelenMamontova committed:
The strings server.cpp changed to main.cpp, server.o changed to main.o,
server changed to main.

HelenMamontova committed:
The tcpserver.h file renamed to server.h.

HelenMamontova committed:
TCPSERVER_H changed to SERVER_H tn #ifndef, #define. The Tcpserver class
name changed to Server. The constructor name Tcpserver changed to
Server.

HelenMamontova committed:
"tcpserver.h" header file replaced by "server.h".

HelenMamontova committed:
"tcpserver.h" header file replaced by "server.h" in tcpserver.cpp file.

HelenMamontova committed:
Constructor call Tcpserver replaced by a call to the constructor Server in main.cpp.

HelenMamontova committed:
Tcpserver class name and Tcpserver constructor name replaced by Server
in tcpserver.cpp.

HelenMamontova committed:
tcpserver.cpp file renamed to server.cpp.

HelenMamontova committed:
The tcpserver.cpp string changed to server.cpp, tcpserver.o string changed to
server.o.

HelenMamontova committed:
The tcpconnection.h file renamed to connection.h.

HelenMamontova committed:
TCPCONNECTION_H changed to CONNECTION_H tn #ifndef, #define. The Tcpconnection class
    name changed to Connection. The constructor name Tcpconnection changed to
    Connection.

HelenMamontova committed:
The std:shared_ptr<Tcpconnection> type of the 'connection' parameter
changed to std::shared_ptr<Connection> in handle_accept() method.

HelenMamontova committed:
The parameter 'new Tcpconnection' changed to new Connection' in
start_accept() method.

HelenMamontova committed:
Tcpconnection class name and Tcpconnection constructor name replaced by Connection in tcpconnection.cpp.

HelenMamontova committed:
The tcpconnection.cpp file renamed to connection.cpp.

HelenMamontova committed:
The tcpconnection.cpp string changed to connection.cpp, tcpconnection.o string changed to
connection.o in Makefile.

HelenMamontova committed:
"tcpconnection.h" header file replaced by "connection.h" in server.h file.

HelenMamontova committed:
"tcpconnection.h" header file replaced by "connection.h" in server.cpp file.

HelenMamontova committed:
"tcpconnection.h" header file replaced by "connection.h" in connection.cpp file.

HelenMamontova committed:
The main string changed to server.

HelenMamontova committed:
The main string changed to server.

HelenMamontova committed:
The 'try' and 'catch' blocks added.

HelenMamontova committed:
The "connection.h" header file removed.

HelenMamontova committed:
The read_file() method definition added.

HelenMamontova committed:
The <dirent.h> header file removed. The read_file() method definition
and call added.

HelenMamontova committed:
Check of the return value of the read() function added to the read_file() method.

HelenMamontova committed:
Class member variable std:vecto<char> msg_ added.

HelenMamontova committed:
Error message fixed in the handle_write() function. The 'msg' variable
changed to 'msg_' class member variable in the handle_read() function.

HelenMamontova committed:
The std::string& work_dir parameter of the Server() constructor replaced by
const std::string& work_dir.

HelenMamontova committed:
The std::string& work_dir parameter of the Server() constructor replaced by const std::string& work_dir.

HelenMamontova committed:
The std::string& work_dir parameter of the Connection() constructor and
of the make_response() method replaced by const std::string& work_dir.
The std::string& str parameter of the string_to_vector_char() method
replaced by const std::string& str.

HelenMamontova committed:
The std::string& work_dir parameter of the Connection() constructor and
of the make_response() method replaced by const std::string& work_dir.
The std::string& str parameter of the string_to_vector_char() method
replaced by const std::string& str.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+    return found ? 0 : 1;
+}
+
+std::vector<char> Connection::string_to_vector_char(const std::string str)
```
I propose the following:
```c++
using Data = std::vector<char>;
// ...
Data toData(const std::string& source)
{
    return Data(source.begin(), source.end());
}
```

Also, you missed `&` in your function.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+}
+
+
+std::vector<char> Connection::read_file(const Request& request, const std::string& path)
```
From the whole `Request` you only need the `path` part. Consider the following:
```c++
Data Connection::read_file(const std::string& path)
{
    // ...
}
```
And at the point of call:
```c++
read_file(work_dir_ + "/" + request.path());
```

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+        else if (errno == EACCES)
+            return string_to_vector_char("HTTP/1.1 403 File access not allowed\r\nContent-Type: text/plain\r\n\r\n403 File access not allowed.\n");
+        else
+            return string_to_vector_char("HTTP/1.1 500 File open error\r\nContent-Type: text/plain\r\n\r\n500 File open error." + std::string(strerror(errno)) + "\n");
```
There is a repeating pattern in these error answers:
```
HTTP/1.1 <code> <title>\r\n
Content-Type: text/plain\r\n\r\n
<message>
```
Only the `<code>`, `<title>` and `<message>` change. So consider making a function:
```c++
Data makeError(unsigned code, const std::string& title, const std::string& message)
{
    return toData("HTTP/1.1 " + std::to_string(code) + " " + title + "\r\nContent-Type: text/plain\r\n\r\n" + message);
}
```
Ans call it like this:
```c++
if (errno = ENOENT)
    return makeError(404, "File does not exist", "404 File does not exist");
```

Also, it would be nice to add file path to the message, so the client know which file does not exist.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+            return string_to_vector_char("HTTP/1.1 500 File open error\r\nContent-Type: text/plain\r\n\r\n500 File open error." + std::string(strerror(errno)) + "\n");
+    }
+
+    std::string ext = request.path().substr(request.path().find(".") + 1);
```
In case of the 'my.beautiful.flower.jpg' filename you get 'beautiful.flower.jpg' as extension. You need to look for the last '.' symbol, not the first.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+
+    std::string ext = request.path().substr(request.path().find(".") + 1);
+
+    for (size_t i = 0; i < ext.length(); i++)
```
Consider making it a function. It does not repeat, but it is a distinct block of code.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+    if (request.version() != "HTTP/1.1" && request.version() != "HTTP/1.0")
+        return string_to_vector_char("HTTP/1.1 505 HTTP Version Not Supported\r\nContent-Type: text/plain\r\n\r\n505 HTTP Version Not Supported.\n");
+
+    const std::string path = work_dir_.empty() ? "." : work_dir_;
```
You can do it in the constructor.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+
+    struct stat st;
+    if (stat((path + "/" + request.path()).c_str(), &st) < 0)
+        return string_to_vector_char("HTTP/1.1 404 File does not exist\r\nContent-Type: text/plain\r\n\r\n404 File does not exist.\n");
```
In this case your `fd` is not closed.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+    if (read(fd, buff.data(), st.st_size) >= 0)
+    {
+        buff.insert(buff.begin(), header.begin(), header.end());
+        return buff;
```
You need to close the `fd`

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    const std::string buff_str(buff_, bytes);
+    message_ += buff_str;
```
Consider `message_.append(buff_, bytes);`

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+        const std::string start_str = message_.substr(0, str_end_pos);
+        msg_ = make_response(Request(start_str), work_dir_);
+
+        boost::asio::async_write(socket_, boost::asio::buffer(msg_, msg_.size()),
```
I think `boost::asio::buffer(msg_)` is enough.

madf commented on connection.h:
```diff
@@ -0,0 +1,34 @@
+        char buff_[1024];
+        std::string message_;
+        std::string work_dir_;
+        std::vector<char> msg_;
```
I think `response_` is a better name.

madf commented on server.cpp:
```diff
@@ -0,0 +1,106 @@
+{
+    namespace pls = std::placeholders;
+
+    resolver_.async_resolve(tcp::resolver::query(host, port), bind(&Server::handle_resolve, this, pls::_1, pls::_2));
```
Which `bind` do you use? `boost::bind` or `std::bind`?

madf commented on server.cpp:
```diff
@@ -0,0 +1,106 @@
+void Server::start_accept()
+{
+     connection_ptr connection(new Connection(io_service_, work_dir_));
+     acceptor_.async_accept(connection->socket(), bind(&Server::handle_accept, this, connection, boost::asio::placeholders::error));
```
Please use `std::bind` and `std::placeholders::_*`

madf commented on server.cpp:
```diff
@@ -0,0 +1,106 @@
+                }
+                catch (const std::exception& e)
+                {
+                    acceptor_.close();
```
Is the `acceptor_` open here? Do we need to close it?

madf commented on server.cpp:
```diff
@@ -0,0 +1,106 @@
+    resolver_.async_resolve(tcp::resolver::query(host, port), bind(&Server::handle_resolve, this, pls::_1, pls::_2));
+}
+
+std::string Server::make_daytime_string()
```
This function does not use anything from `Server`. Consider making it a free function.

madf commented on server.cpp:
```diff
@@ -0,0 +1,106 @@
+    return buffer;
+}
+
+std::string Server::make_log_line(const std::string& message)
```
Like the previous one, this function does not use anything from `Server`. Make it free.

madf commented on server.cpp:
```diff
@@ -0,0 +1,106 @@
+{
+    if (!err)
+    {
+        try
```
What does this `try` catch?

madf commented on server.h:
```diff
@@ -0,0 +1,29 @@
+        boost::asio::ip::tcp::acceptor acceptor_;
+        std::string outfile_;
+        std::string work_dir_;
+        char buff_[1024];
```
Where the `buff_` is used?

HelenMamontova committed:
Using-statement using Data added. 'String_to_vector_char' function name changed to 'toData'.
Parameter const std:string str of the function toData() changed to const
std::string& source.

HelenMamontova committed:
'String_to_vector_char' function name changed to 'toData'.
Parameter const std:string str of the function toData() changed to const std::string& source.
Unnecessary variable 'buffer' removed in function toData().

HelenMamontova committed:
Std::vector<char> type of functions read_file(), make_index(), make_response()
and variable msg_ changed to Data.

HelenMamontova committed:
Std::vector<char> type of functions read_file(), make_index(), make_response()
and variables index, buff changed to Data.

HelenMamontova committed:
'const Request& request' parameter of read_file() function removed.

HelenMamontova committed:
Parameter 'const Request& request' removed, parameter of the function open() changed,
'std::string ext' variable definition changed, the file name in the stat() function replaced to 'path'
in the definition of the function read_file(). The parameter of the read_file() function at the call point changed.

HelenMamontova committed:
The makeError() method declaration added.

HelenMamontova committed:
The makeError() method definition added. The makeError() method call
added to the read_file() function.

HelenMamontova committed:
The 'path' parameter removed in the makeError() function.

HelenMamontova committed:
The 'path' parameter removed in the makeError() function. The 'path' added
to the 'message' parameter when calling makeError() function in the read_file() function.

HelenMamontova committed:
The string_to_lower() function declaration added.

HelenMamontova committed:
The string_to_lower() function definition and call added.
The close(fd) call added where necessary.

HelenMamontova committed:
The 'path' parameter removed in the make_index() function. The
'work_dir_' parameter removed in the make_response() function.

HelenMamontova committed:
The 'path' variable definition removed in the make_response() function. The
'work_dir_' parameter removed in the make_response() function. The
'path' variable replaced by 'work_dir_' variable. The 'work_dir_'
variable definition added in the Connection() constructor.

HelenMamontova committed:
The 'toData' and 'makeError' function names changed to 'to_data' and
'make_error' names.

HelenMamontova committed:
The 'toData' and 'makeError' function names changed to 'to_data' and 'make_error' names.

HelenMamontova committed:
Unnecessary variable buff_str removed. Definition of variable message_ changed - added function call append().

HelenMamontova committed:
In the handle_read() function boost::asio::buffer(buff_, buff_.size()) parameter
of the async_write() function changed to boost::asio::buffer(buff_).

HelenMamontova committed:
The Data msg_ variable name changed to Data response_.

HelenMamontova committed:
The Data msg_ variable name changed to Data response_.

HelenMamontova committed:
The <boost/bind.hpp> header file changed to <functional>. The namespace
pls moved from Server() constructor. Boost::asio::placeholders::error
changed to pls::_1 in the handle_accept() function.

HelenMamontova committed:
The make_daytime_string() function moved from the class Server.

HelenMamontova committed:
The make_daytime_string() function declaration removed.

HelenMamontova committed:
The make_log_line() function declaration removed.

HelenMamontova committed:
The make_log_line() function definition moved from the class Server.

HelenMamontova committed:
Block try-catch removed outside the cycle while in function
handle_resolve().

HelenMamontova committed:
The acceptor_.close() removed in the handle_resolve() function.

HelenMamontova committed:
The unnecessary variable buff_ removed.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,203 @@
+    return found ? 0 : 1;
+}
+
+Data Connection::to_data(const std::string& source)
```
This function does not use private data members, so it should not be a member. Make it a free function.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,203 @@
+            close(fd);
+            return make_error(403, "File access not allowed", path + ": 403 File access not allowed.");
+        }
+        else
```
You don't need 'else' - you return from 'if'.

madf replied:
I mean, instead of:
```c++
if (somethig)
{
    do_something();
    return;
}
else
{
    do_something_else();
    return;
}
```
you can do:
```c++
if (somethig)
{
    do_something();
    return;
}

do_something_else();
return;
```

madf commented on connection.cpp:
```diff
@@ -0,0 +1,203 @@
+        header = "HTTP/1.1 200 OK\r\nContent-Type: application/octet-stream\r\nContent-Disposition: attachment\r\n\r\n";
+
+    struct stat st;
+    if (stat(path.c_str(), &st) < 0)
```
You may do this call before opening the file so you won't need to close it.

madf replied:
How about this?

madf commented on connection.cpp:
```diff
@@ -0,0 +1,203 @@
+    }
+    else
+    {
+        DIR *dir = opendir(work_dir_.c_str());
```
I think you can safely move `opendir` and `closedir` inside the `make_index`.

madf replied:
How about this?

madf commented on connection.cpp:
```diff
@@ -0,0 +1,203 @@
+        response_ = make_response(Request(start_str));
+
+        boost::asio::async_write(socket_, boost::asio::buffer(response_),
+        boost::asio::transfer_all(),
```
Something is wrong with indentation here.

madf commented on main.cpp:
```diff
@@ -0,0 +1,99 @@
+    const std::string version = "1.0.0";
+    std::string address;
+    std::string outfile;
+    std::string work_dir;
```
I think you can initialize `work_dir` with '.' so you don't need to check it for emptiness elsewhere.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,203 @@
+void Connection::start()
+{
+    boost::asio::async_read(socket_, boost::asio::buffer(buff_),
+        bind(&Connection::read_complete, shared_from_this(), pls::_1, pls::_2),
```
Which `bind` do you use: `std::bind` or `boost::bind`?

HelenMamontova committed:
The unnecessary 'else' removed in the if(errno==EACCES) construction in
read_file() function.

HelenMamontova committed:
The std:: prefix added to bind() function.

HelenMamontova committed:
The variable work_dir assigned the value '.' when variable declared.

HelenMamontova committed:
The work_dir_ variable initialization added in the constructor
Connection(). Connection() constructor body removed.

HelenMamontova committed:
The to_data() function declaration removed.

HelenMamontova committed:
The to_data() function definition moved from the class Connection. The
'using Data = std::vector<char>' added.

HelenMamontova committed:
The function make_error() declaration removed.

HelenMamontova committed:
The function make_error() definition moved from the class Connection.

HelenMamontova committed:
The function string_to_lower() declaration removed.

HelenMamontova committed:
The function string_to_lower() definition moved from the class Connection.

HelenMamontova committed:
The function read_file() declaration removed.

HelenMamontova committed:
The function read_file() definition moved from the class Connection.

madf commented on connection.cpp:
```diff
@@ -9,41 +9,28 @@
 }
 
-std::string Connection::string_to_lower(std::string str)
+std::string string_to_lower(std::string str)
```
`to_lower` is better.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,201 @@
+    {
+        if (errno == EACCES)
+        {
+            close(fd);
```
You may close the `fd` before `if`, so you don't need to close it in two places.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,201 @@
+Data Connection::make_response(const Request& request)
+{
+    if (request.verb() != "GET")
+        return to_data("HTTP/1.1 405 Method not allowed\r\nContent-Type: text/plain\r\n\r\n405 Method not allowed.\n");
```
`make_error`?

HelenMamontova committed:
The string_to_lower() function renamed to to_lower().

HelenMamontova committed:
The code 'struct stat st; if(stat(path.c_str(), &st < 0)' moved before opening the file code
in the read_file() function.

HelenMamontova committed:
The code 'close(fd)' moved before code 'if(errno == EACCES)' in the read_file() function.

HelenMamontova committed:
The return of the error message  replaced by a function call make_error()
in the make_response() function.

HelenMamontova committed:
The DIR*dir paremeter removed in the makeindex() function.

HelenMamontova committed:
The DIR*dir paremeter removed from the definition and call make_index() function.
DIR* dir = opendir() and close(dir) moved into the make_index()
function.

madf commented on server.cpp:
```diff
@@ -0,0 +1,98 @@
+        outfile_(outfile),
+        work_dir_(work_dir)
+{
+    resolver_.async_resolve(tcp::resolver::query(host, port), bind(&Server::handle_resolve, this, pls::_1, pls::_2));
```
What kind of `bind`? `boost::bind` or `std::bind`?

HelenMamontova committed:
The std:: prefix added to bind() function in the Server() constructor
and start_accept() function.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+        close(fd);
+        if (errno == EACCES)
+            return make_error(403, "File access not allowed", path + ": 403 File access not allowed.");
+        return make_error(500, "The file descriptor is invalid.", path + ": 500 The file descriptor is invalid." + std::string(strerror(errno)));
```
I don't think 'The file descriptor is invalid' is a correct message here. First, on the client side there are no file descriptors, this message does not tell anything to the client. Second, there is a wide range of different errors that may happen to `read`, and only one of them is about invalid file descriptor. So I propose to change the wording to more standard-conformant 'Internal server error'.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+    {
+        close(fd);
+        if (errno == EACCES)
+            return make_error(403, "File access not allowed", path + ": 403 File access not allowed.");
```
'File access **is** not allowed'

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+        if (errno == ENOENT)
+            return make_error(404, "File does not exist", path + ": 404 File does not exist.");
+        else if (errno == EACCES)
+            return make_error(403, "File access not allowed", path + ": 403 File access not allowed.");
```
'File access **is** not allowed'

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+        else if (errno == EACCES)
+            return make_error(403, "File access not allowed", path + ": 403 File access not allowed.");
+        else
+            return make_error(500, "File open error", path + ": 500 File open error." + std::string(strerror(errno)));
```
Let's make it 'Internal server error', like below.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+{
+    DIR *dir = opendir(work_dir_.c_str());
+    if (dir == NULL)
+        return make_error(500, "Failed to open directory", "500 Failed to open directory.");
```
It would be nice to have the exact reason of the error, like you do it with files using `strerror(errno)`

madf commented on connection.cpp:
```diff
@@ -0,0 +1,191 @@
+    }
+    else
+    {
+        boost::asio::async_read(socket_, boost::asio::buffer(buff_),
```
This call is identical to `start` method. I think you may call `start` here instead.

madf commented on request.cpp:
```diff
@@ -0,0 +1,28 @@
+
+Request::Request(const std::string& start_str)
+{
+    const size_t pos_space_first = start_str.find_first_of(' ');
```
What if there is no ' ' symbol?

HelenMamontova committed:
Error message fixed in the construction if(errno==EACCES) in the
function read_file().

HelenMamontova committed:
Error message with code 500 fixed in the function read_file().

HelenMamontova committed:
Error message with code 403 fixed in the construction if(read()>=0) in the function read_file().

HelenMamontova committed:
Error message with code 500 fixed in the construction if(read()>=0) in the function read_file().

HelenMamontova committed:
The std::string(strerror(errno)) added to error message with code 500
in the construction if(dirr==NULL) in the function make_index().

HelenMamontova committed:
The call async_read() replaced by call start() in the handle_read()
function.

HelenMamontova committed:
The header file "request.h removed.

HelenMamontova committed:
The pos_space_first and pos_space_second variables checking added in the
Request() constructor.

HelenMamontova committed:
In the function handle_read() added blocks try and catch,
added async_write() call in the catch block.

HelenMamontova committed:
The new file log.h added.

HelenMamontova committed:
The new file log.cpp added.

HelenMamontova committed:
The files log.cpp, log.o added.

HelenMamontova committed:
The write_log() function declaration removed.

HelenMamontova committed:
The functions write_log(), make_daytime_string(), make_log_line()
definition removed. The 'outfile_ parameter added when calling the
write_log() function. The "log.h" header file added.

HelenMamontova committed:
The outfile_ parameter added when creating a connection object.

HelenMamontova committed:
The outfile parameter added to the Connection() constructor. The member
of class variable outfile_ added.

HelenMamontova committed:
The "log.h" header file added. The outfile parameter added in the
Connection() constructor. Initialization of the outfile_ variable
in the Connection() constructor added. The write_log() function call
added in the handle_read() function.

madf commented on connection.cpp:
```diff
@@ -42,9 +43,9 @@ Data read_file(const std::string& path)
+            return make_error(403, "File access is not allowed", path + ": 403 File access is not allowed.");
         else
-            return make_error(500, "File open error", path + ": 500 File open error." + std::string(strerror(errno)));
+            return make_error(500, "Internal server error", path + ": 500 Internal server error." + std::string(strerror(errno)));
```
Please add a whitespace after '.'.

madf commented on connection.cpp:
```diff
@@ -66,14 +67,15 @@ Data read_file(const std::string& path)
-            return make_error(403, "File access not allowed", path + ": 403 File access not allowed.");
-        return make_error(500, "The file descriptor is invalid.", path + ": 500 The file descriptor is invalid." + std::string(strerror(errno)));
+            return make_error(403, "File access is not allowed", path + ": 403 File access is not allowed.");
+        return make_error(500, "Internal server error.", path + ": 500 Internal server error." + std::string(strerror(errno)));
```
Please add a whitespace after the '.'.

madf commented on connection.cpp:
```diff
@@ -89,7 +91,7 @@ Data Connection::make_index()
     DIR *dir = opendir(work_dir_.c_str());
     if (dir == NULL)
-        return make_error(500, "Failed to open directory", "500 Failed to open directory.");
+        return make_error(500, "Failed to open directory", "500 Failed to open directory." + std::string(strerror(errno)));
```
Please add a whitespace after the '.'.

madf commented on request.cpp:
```diff
@@ -3,10 +3,14 @@
 {
     const size_t pos_space_first = start_str.find_first_of(' ');
+    if (pos_space_first == std::string::npos)
+        throw "Bad request.";
```
Do not use plain values for throwing exceptions. Use `std::exception` and its derivatives - `std::runtime_error`, `std::logic_error` or whatever is more appropriate.

madf commented on connection.cpp:
```diff
@@ -171,17 +173,26 @@ void Connection::handle_read(const error_code& error, size_t bytes)
+        }
+        catch (const char* exception)
+        {
+            std::cerr << "Exception: " << exception << "\n";
```
What is the purpose of `std::cerr` here?

madf commented on connection.cpp:
```diff
@@ -171,17 +173,26 @@ void Connection::handle_read(const error_code& error, size_t bytes)
+        catch (const char* exception)
+        {
+            std::cerr << "Exception: " << exception << "\n";
+            write_log(socket().remote_endpoint().address().to_string() + " 400 Bad request", outfile_);
```
Since you are logging this error, it would be nice to log all other errors.

HelenMamontova committed:
The spaces added to error messages in make_error() function call in the
read_file) function.

HelenMamontova committed:
The space added to error message in make_error() function call in the make_index() function.

HelenMamontova committed:
The std::cerr removed in the block catch in the handle_read() function.

HelenMamontova committed:
The write_log() function call added when checking verb and version.

HelenMamontova committed:
The write_log() function call added when checking dir in the
make_index() function.

HelenMamontova committed:
The new file error.h added.

HelenMamontova committed:
The new file error.cpp added.

HelenMamontova committed:
The error.cpp, error.o files added in the Makefile.

HelenMamontova committed:
Header file "error.h added. In the function read_file() in the construction if(stat() < )
return make_error() replaced by throw Error(errno, path).
In the function handle_read()  the block catch(const char* exceptin) is commented out,
the catch(const Error& exception) block is added.

HelenMamontova committed:
In the function read_file() in the construction if(fd=-1) errno checks replaced by
throw Error(errno, path). In the function handle_read() the checks 'if(code==EACCES)'
and 'else' added to the catch(const Error& exception) block.

HelenMamontova committed:
In the function read_file() in the construction if(read >= 0)-else errno checks replaced by throw Error(errno, path).

HelenMamontova committed:
In the function make_index() in the construction if(dir == NULL) write_log() and
make_error() call functions replaced by throw Error(errno, work_dir_).

HelenMamontova committed:
The "std::cout" changed to write_log() call in the handle_write() and
handle_read() functions.

HelenMamontova committed:
Header file <exception> added, block catch(const char* exception) replaced
by block catch(std::exception &exception), formatting fixed, the
std::string msg variable added in the handle_read function.

HelenMamontova committed:
Header file <stdexcept> added. Exception in operator throw changed in
the Request() constructor.

HelenMamontova committed:
The 'path added to the 'message' parameter of write_log() function call
in the handle_read() function.

HelenMamontova committed:
The header files "badverb.h" and "badversion.h" added.
In the if() block, when checking verb, the code replaced by throw Badverb(),
in the if() block, when checking version, the code replaced by throw
Badversion() in the function make_response(). In the handle_read()
function the blocks catch(const Badverb &exception) and catch(const
Badversion &exception) added.

HelenMamontova committed:
The new file badverb.h added.

HelenMamontova committed:
The new file badversion.h added.

HelenMamontova committed:
The constructor and methods definition added.

HelenMamontova committed:
The 'error.cpp' and 'error.o' notes removed.

HelenMamontova committed:
The error.cpp file removed.

HelenMamontova committed:
In the handle_accept() function 'std::cout' changed to write_log()
function.

HelenMamontova committed:
The "std::cout" changed to write_log() call in the handle_reslve()
function.

HelenMamontova committed:
The constructor Badverb() replaced by 'using
std::runtime_error::runtime_error' in badverb.h file.

HelenMamontova committed:
The constructor Badversion() replaced by 'using std::runtime_error::runtime_error' in badversion.h file.

HelenMamontova committed:
The std::string msg variable renamed to std::string request_error in the
block catch(std::exception &exception) in handle_read() function.

HelenMamontova committed:
The <iostream> header file removed.

HelenMamontova committed:
The <iostream> header file removed. Formatting fixed.

madf commented on badverb.h:
```diff
@@ -0,0 +1,13 @@
+#include <stdexcept>
+#include <string>
+
+class Badverb: public std::runtime_error
```
`Badverb` -> `BadVerb`

madf replied:
Also it makes sense to put this class into `error.h`

madf commented on badversion.h:
```diff
@@ -0,0 +1,13 @@
+#include <stdexcept>
+#include <string>
+
+class Badversion: public std::runtime_error
```
`Badversion` -> `BadVersion`. Also please put it into `error.h`

madf commented on connection.cpp:
```diff
@@ -132,10 +126,10 @@ Data Connection::make_index()
 {
     if (request.verb() != "GET")
-        return make_error(405, "Method not allowed", "405 Method not allowed.");
+        throw Badverb(std::string("Method not allowed."));
```
Why `std::string`? I think you can construct it like this: `Badverb("Method not allowed.")`

madf commented on connection.cpp:
```diff
@@ -1,9 +1,12 @@
 #include <fstream>
 #include <vector>
 #include <algorithm>
+#include <exception>
```
Do you need this header?

madf commented on connection.cpp:
```diff
@@ -181,11 +175,53 @@ void Connection::handle_read(const error_code& error, size_t bytes)
+        catch (const Error &exception)
+        {
+            int code = exception.code();
+            std::string path = exception.path();
```
What is the reason of having `code` and `path` variables? Can you use `exception.code()` and `exception.path()` instead?

madf commented on connection.cpp:
```diff
@@ -181,11 +175,53 @@ void Connection::handle_read(const error_code& error, size_t bytes)
+                boost::asio::async_write(socket_, boost::asio::buffer(make_error(500, "Internal server error", path + ": 500 Internal server error. " + std::string(strerror(code)))),
+                    boost::asio::transfer_all(),
+                    std::bind(&Connection::handle_write, shared_from_this(), pls::_1, pls::_2));
+            }
```
These three `if` branches are almost identical. Please put them into a function. Also, consider using `switch/case` instead of `if`.

madf commented on connection.cpp:
```diff
@@ -181,11 +175,53 @@ void Connection::handle_read(const error_code& error, size_t bytes)
+        }
+        catch (const Badverb &exception)
+        {
+            std::string verb_error(exception.what());
```
Do you need this variable?

madf commented on connection.cpp:
```diff
@@ -181,11 +175,53 @@ void Connection::handle_read(const error_code& error, size_t bytes)
+                boost::asio::transfer_all(),
+                std::bind(&Connection::handle_write, shared_from_this(), pls::_1, pls::_2));
+        }
+        catch (std::exception &exception)
```
Please add `const`

madf commented on connection.cpp:
```diff
@@ -181,11 +175,53 @@ void Connection::handle_read(const error_code& error, size_t bytes)
+            write_log(socket().remote_endpoint().address().to_string() + " 400 " + request_error, outfile_);
+           boost::asio::async_write(socket_, boost::asio::buffer(make_error(400, request_error, "400 " + request_error)),
                 boost::asio::transfer_all(),
                 std::bind(&Connection::handle_write, shared_from_this(), pls::_1, pls::_2));
```
From where this exception can come from? If it's a 'just in case' scenario, '500 Internal Server Error' looks more appropriate.

madf commented on error.h:
```diff
@@ -0,0 +1,32 @@
+            return code_;
+        }
+
+        std::string path() const
```
`const std::string&` would be better.

madf commented on request.cpp:
```diff
@@ -1,16 +1,17 @@
     const size_t pos_space_first = start_str.find_first_of(' ');
     if (pos_space_first == std::string::npos)
-        throw "Bad request.";
+        throw std::runtime_error("Bad request.");
```
Why `std::runtime_error`? You have `Error`, `BadVerb`, `BadVersion` - so maybe add another special class for these errors?

madf commented on server.cpp:
```diff
@@ -33,7 +32,7 @@ void Server::handle_accept(connection_ptr connection, const error_code& error)
     else
     {
-        std::cout << "Error async_accept: " << error.message() << "\n";
+        write_log(connection->socket().remote_endpoint().address().to_string() + "Error async_accept: " + error.message(), outfile_);
```
I think, in case of `async_accept` error you don't have a connection, so `remote_endpoint` does not exist at this moment.

HelenMamontova committed:
The class name Getverb changed to GetVerb in getverb.h file.

HelenMamontova committed:
The class name Getversion changed to GetVersion in badversion.h file.

HelenMamontova committed:
The name of constructors Badverb and Badversion changed to BadVerb and BadVersion in connection.cpp file.

HelenMamontova committed:
Definition of BadVerb and BadVersion classes moved to error.h file.

HelenMamontova committed:
The header files "badverb.h" and "badversion.h" removed from
connection.cpp file.

HelenMamontova committed:
The files badverb.h and badversion.h removed.

HelenMamontova committed:
The header file <exception> removed from connection.cpp file.

HelenMamontova committed:
The "std::string" removed in the operator throw.

HelenMamontova committed:
Unnecessary variables code and path removed in the handle_read()
function.

HelenMamontova committed:
Unnecessary variables verb_error and version_error removed in the handle_read() function.

HelenMamontova committed:
The 'const' added in the block catch(std::exception &exception).

HelenMamontova committed:
Class BadRequest added in error.h file.

HelenMamontova committed:
The <stdexcept> header file removed, "error.h" header file added.
Operatr throw std::runtime_error() changed to operator throw
BadRequest() in the constructor Request().

HelenMamontova committed:
In the block catch 'std::exception &exception' changed to 'const
BadRequest &exception' in the handle_read() function.

HelenMamontova committed:
The type of std::string path() method changed to const std::string& in
the error.h file..

HelenMamontova committed:
In the function handle_accept(), in case of an error, the address removed from the first parameter of write_log() function.

HelenMamontova committed:
Unnecessary variable request_error removed in the function handle_read().

HelenMamontova committed:
The handle_exception() method declaration added.

HelenMamontova committed:
In the connection.cpp file the handle_exception() method definition added, handle_exception()
method call added in block catch(const Error &exception) in the
handle_read() function.

HelenMamontova committed:
In connection.cpp file the handle_exception() method call added
in the block catch(const BadVerb &exception) in the handle_read() function.

HelenMamontova committed:
In connection.cpp file the handle_exception() method call added
in the block catch(const BadVersion &exception) in the handle_read() function.

HelenMamontova committed:
In connection.cpp file the handle_exception() method call added
in the block catch(const BadRequest &exception) in the handle_read() function.

HelenMamontova committed:
The if/else operator replaced by operator switch
in the catch(const Error &exception) block in the handle_read() function.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,220 @@
+    else
+    {
+        close(fd);
+        throw Error(errno, path);
```
At this point `errno` will probably be overwritten by `close`. I mean, if we get,say, `EINTR` in the `read` call, the `close` call will be successful, and `errno` will be set to 0.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,220 @@
+{
+    if (error)
+    {
+        write_log(socket().remote_endpoint().address().to_string() + " Error async_write: " + error.message(), outfile_);
```
You can access `socket_` directly here. No need to call `socket()`.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,220 @@
+{
+    if (error)
+    {
+        write_log(socket().remote_endpoint().address().to_string() + " Error async_read: " + error.message(), outfile_);
```
I see that `socket().remote_endpoint().address().to_string()` repeats. How about making an overloaded version of `write_log` like the following:
```c++
void write_log(const socket_type& socket, const std::string& message, const std::string& outFile)
{
    write_log(socket.remote_endpoint().address().to_string() + " " + message, outFile);
}
```
and use it like `write_log(socket_, " Foo bar baz", outfile_)`.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,220 @@
+
+using boost::asio::ip::tcp;
+using boost::system::error_code;
+using Data = std::vector<char>;
```
You have the same declaration in the header.

madf commented on error.h:
```diff
@@ -0,0 +1,50 @@
+{
+    public:
+        Error(int code, const std::string& path) noexcept
+            : runtime_error(std::to_string(code)),
```
Bad idea. If you accidentally or intentionally catch this error by its base class (e.g. `std::exception`), the `what()` member will return hust a tring representation of the error code. "-15" does not say anything, does it? It's better to turn the code into a real error message using `strerror`.

madf commented on error.h:
```diff
@@ -0,0 +1,50 @@
+        std::string path_;
+};
+
+class BadRequest: public std::runtime_error
```
If you use `struct` you can shorten this and other such classes to just:
```c++
struct BadRequest : std::runtime_error
{
    using std::runtime_error::runtime_error;
};
```

madf commented on request.cpp:
```diff
@@ -0,0 +1,33 @@
+{
+    const size_t pos_space_first = start_str.find_first_of(' ');
+    if (pos_space_first == std::string::npos)
+        throw BadRequest("Bad request.");
```
I think it may be useful to be more verbose here. To say why the request is bad, actually. The same below.

madf commented on server.cpp:
```diff
@@ -0,0 +1,75 @@
+
+void Server::start_accept()
+{
+     connection_ptr connection(new Connection(io_service_, work_dir_, outfile_));
```
Avoid using `new` with smart pointers. In this case consider using `std::make_shared`.

madf commented on server.cpp:
```diff
@@ -0,0 +1,75 @@
+        tcp::resolver::iterator end;
+        while (endpoint_iterator != end)
+        {
+            tcp::endpoint ep = *endpoint_iterator;
```
`const`?

madf commented on server.cpp:
```diff
@@ -0,0 +1,75 @@
+{
+    if (!err)
+    {
+        tcp::resolver::iterator end;
```
`const`?

madf commented on log.cpp:
```diff
@@ -0,0 +1,26 @@
+std::string make_daytime_string()
+{
+    char buffer[80];
+    time_t now = time(nullptr);
```
`const`?

madf commented on connection.cpp:
```diff
@@ -0,0 +1,220 @@
+    if (fd == -1)
+        throw Error(errno, path);
+
+    std::string ext = to_lower(path.substr(path.rfind(".") + 1));
```
`const`?

madf commented on connection.cpp:
```diff
@@ -0,0 +1,220 @@
+                boost::asio::transfer_all(),
+                std::bind(&Connection::handle_write, shared_from_this(), pls::_1, pls::_2));
+        }
+        catch (const Error &exception)
```
In other places you put a whitespace after the ampersand, like this: `const Type& value`. Be consistent. See below as well.

HelenMamontova committed:
The unnecessary using Data = std::vector <char> deleted in
connection.cpp.

HelenMamontova committed:
The 'const' added to std::string ext variable.

HelenMamontova committed:
The 'err' variable added in the read_file() function.

HelenMamontova committed:
<boost/asio.hpp> header file and overloaded function write_log() added
in log.h file.

HelenMamontova committed:
The definition of the overloaded function write_log() added in log.cpp file.

HelenMamontova committed:
The 'socket().remote_endpoint().address().to_string()' parameter
is replaced by 'socket_' in the write_log() function.

HelenMamontova committed:
Spaces formatting fixed in the handle_read() function.

HelenMamontova committed:
The header file <cstring> added. The std::to_string(code) parameter of
runtime_error() changed to strerror(code) in the constructor Error().

HelenMamontova committed:
Class BadRequest replaced by struct BadRequest, class BedVerb replaced
by struct BadVerb, class BadVersion replaced by struct BadVersion.

HelenMamontova committed:
The 'const' added to variable time_t now.

HelenMamontova committed:
Messages in throw BadRequest() changed to more detailed.

HelenMamontova committed:
In the start_accept() function 'new Connection()' changed to
'std::make_shared<Connection>'.

HelenMamontova committed:
The 'const' added to variables tcp::resolver::iterator end,
tcp::endpoint ep in the handle_resolve() function.

madf commented on connection.h:
```diff
@@ -0,0 +1,36 @@
+using Data = std::vector<char>;
+
+class Connection
+    : public std::enable_shared_from_this<Connection>
```
`std::enable_shared_from_this` comes from the `<memory>` header, so you need to include it.

madf commented on connection.cpp:
```diff
@@ -0,0 +1,220 @@
+
+    message_.append(buff_, bytes);
+
+    if (bytes < 1024)
```
It's better to reorganize this condition a bit:
```c++
if (bytes == 1024)
{
    start();
    return;
}

// All other code
```

madf commented on server.cpp:
```diff
@@ -0,0 +1,75 @@
+
+void Server::start_accept()
+{
+     connection_ptr connection(std::make_shared<Connection>(io_service_, work_dir_, outfile_));
```
```c++
connection_ptr connection = std::make_shared<Connection>(io_service_, work_dir_, outfile_);
```
 - a little bit more idiomatic.

HelenMamontova committed:
The header file <memory> added in connection.h file.

HelenMamontova committed:
The condition if(bytes<1024) changed to if(bytes==1024) and
corresponding code moved from if() in the handle_read() function.

HelenMamontova committed:
The way to initialize connection_ptr connection change from parentheses () to = sign
in the start_accept() function.
